wave_optimization/include/wave/optimization/gtsam/pose_vel.hpp

```diff
@@ -0,0 +1,230 @@
+            retval.pose = traits<Pose3>::Expmap(v.block<6, 1>(0, 0)., J1);
+            retval.vel =
+              traits<decltype(m.vel)>::Logmap(v.block<6, 1>(6, 0)., J2);
+
+            Hv->block<6, 6>(0, 0).noalias() = *J1;
+            Hv->block<6, 6>(6, 6).noalias() = *J2;
+        } else {
+            retval.pose =
+              traits<Pose3>::Expmap(v.block<6, 1>(0, 0)., boost::none);
+            retval.vel = traits<decltype(m.vel)>::Logmap(v.block<6, 1>(6, 0).,
+                                                         boost::none);
+        }
+        return retval;
+    }
+
+    static wave::PoseVel Compose(const wave::PoseVel &m1,
+                                 const wave::PoseVel &m2,
+                                 ChartJacobian H1 = boost::none,
+                                 ChartJacobian H2 = boost::none) {
+        wave::PoseVel retval;
+        traits<Pose3>::ChartJacobian J1, J2;
+        traits<decltype(m.vel)>::ChartJacobian J3, J4;
+
+        retval.pose = traits<Pose3>::Compose(m1.pose, m2.pose, J1, J2);
+        retval.vel = traits<decltype(m.vel)>::Compose(m1.vel, m2.vel, J3, J4);
+
+        if (H1) {
+            H1->resize(dimension, dimension);
+            H1->setZero();
+            H1->block<6, 6>(0, 0).noalias() = *J1;
+            H1->block<6, 6>(6, 6).noalias() = *J3;
+        }
+        if (H2) {
+            H2->resize(dimension, dimension);
+            H2->setZero();
+            H2->block<6, 6>(0, 0).noalias() = *J2;
+            H2->block<6, 6>(6, 6).noalias() = *J4;
+        }
+        return retval;
+    }
+
+    static wave::PoseVel Between(const wave::PoseVel &m1,
+                                 const wave::PoseVel &m2,
+                                 ChartJacobian H1 = boost::none,
+                                 ChartJacobian H2 = boost::none) {
+        wave::PoseVel retval;
+        traits<Pose3>::ChartJacobian J1, J2;
+        traits<decltype(m.vel)>::ChartJacobian J3, J4;
+
+        retval.pose = traits<Pose3>::Between(m1.pose, m2.pose, J1, J2);
+        retval.vel = traits<decltype(m.vel)>::Between(m1.vel, m2.vel, J3, J4);
+
+        if (H1) {
+            H1->resize(dimension, dimension);
+            H1->setZero();
+            H1->block<6, 6>(0, 0).noalias() = *J1;
+            H1->block<6, 6>(6, 6).noalias() = *J3;
+        }
+        if (H2) {
+            H2->resize(dimension, dimension);
+            H2->setZero();
+            H2->block<6, 6>(0, 0).noalias() = *J2;
+            H2->block<6, 6>(6, 6).noalias() = *J4;
+        }
+        return retval;
+    }
+
+    static wave::PoseVel Inverse(const wave::PoseVel &m,
+                                 ChartJacobian H = boost::none) {
+        wave::PoseVel retval;
+        traits<Pose3>::ChartJacobian J1;
+        traits<decltype(m.vel)>::ChartJacobian J2;
+
+        retval.pose = traits<Pose3>::Inverse(m.pose, J1);
+        retval.vel = traits<decltype(m.vel)>::Inverse(m.vel, J2);
+        if (H) {
```

Comment:
It looks like for all of these operations, you always calculate the Jacobians even when you don't assign to `H`. Could you not calculate `J1` and `J2` if `!H`?

Reply:
Actually, the default constructor of `OptionalJacobian` is like `boost::none`. So maybe you _never_ calculate the Jacobians? Is this tested?

Reply:
I do have a couple tests that optimize on PoseVelBias states which work properly.

Reply:
Suspicious, I don't see how. Make more tests until they break